illumos-utils/src/process.rs

```diff
@@ -0,0 +1,391 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A process executor
+
+use itertools::Itertools;
+use slog::{debug, error, info, Logger};
+use std::os::unix::process::ExitStatusExt;
+use std::process::{Command, ExitStatus, Output};
+use std::str::from_utf8;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+/// Describes the commonly-used "safe-to-reference" type describing the
+/// Executor as a trait object.
+pub type BoxedExecutor = Arc<dyn Executor>;
+
+/// Describes an "executor", which can run [Command]s and return a response.
+///
+/// - In production, this is usually simply a [HostExecutor].
+/// - Under test, this can be customized, and a [FakeExecutor] may be used.
+pub trait Executor: Send + Sync {
+    fn execute(&self, command: &mut Command) -> Result<Output, ExecutionError>;
+}
+
+fn log_command(log: &Logger, id: u64, command: &Command) {
+    info!(log, "{id} - Running Command: [{}]", Input::from(command),);
+    debug!(
+        log,
+        "{id} - Environment: [{}]",
+        to_space_separated_string(command.get_envs()),
+    )
+}
+
+fn log_output(log: &Logger, id: u64, output: &Output) {
+    info!(
+        log,
+        "{id} - {} (status code: {})",
+        if output.status.success() { "OK" } else { "ERROR" },
+        output
+            .status
+            .code()
+            .map(|c| c.to_string())
+            .unwrap_or_else(|| "none".to_string()),
+    );
+    if !output.stdout.is_empty() {
+        debug!(
+            log,
+            "{id} - stdout: {}",
+            from_utf8(&output.stdout).unwrap_or("<Not valid UTF-8>"),
+        );
+    }
+    if !output.stderr.is_empty() {
+        debug!(
+            log,
+            "{id} - stderr: {}",
+            from_utf8(&output.stderr).unwrap_or("<Not valid UTF-8>"),
+        );
+    }
+}
```

Comment:
This is the *other* big advantage of this PR -- under both tests and in production, we'll see the same logging for all commands sent to this executor.